### PR TITLE
GeoNames data augmentation

### PR DESCRIPTION
**Pull request by pvf2005**
This is a data augmentation transformation making use of GeoNames. It will augment sentences containing cities or countries. More information in the README and examples are in test.json.

### CODE REVIEW

File: transformations/geonames_transformation/transformation.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+
+from interfaces.SentenceOperation import SentenceOperation
+from tasks.TaskTypes import TaskType
+
+
+class GeoNamesTransformation(SentenceOperation):
+    tasks = [
+        TaskType.TEXT_CLASSIFICATION,
+        TaskType.TEXT_TO_TEXT_GENERATION,
+        TaskType.TEXT_TAGGING,
```
Comment (ashish3586) [inline transformations/geonames_transformation/transformation.py:13]: Please remove task "TEXT_TAGGING". This transformation is adding new words to a given sentence, so we can't use this for text tagging unless the generation method produces the respective tagging sequence along with the generated transformation.

Comment (pvf2005) [reply transformations/geonames_transformation/transformation.py]: You are right, the method is adding new words. "TEXT_TAGGING" is now removed.

**Comment by kaustubhdhole** [comment]
I feel this is a great idea and can be powerfully used for data augmentation - it is clearly highly precise, and highly accurate from the looks of it, and the coverage is restricted by the names of all locations - and I think that itself covers quite a lot. There is a lot of information in GeoNames and you could make use of almost all of it. 
1. You have a pattern: "COUNTRY with capital" -> you could probably also add the reverse (if the city is a capital city) "city, the capital of COUNTRY"
2. countryInfo gives information about neighbours too - I guess you could add patterns like "COUNTRY, neighbouring COUNTRY2" or "COUNTRY, neighbouring COUNTRY2 & COUNTRY3", etc. or a better pattern something on those lines? I think you could use this to your advantage to amplify the amount of novelty since content on the internet would mostly follow the patterns that you've currently added. I think speaking about neighbouring countries is done and on rare occassions.
3. The country codes could also be used - Romania has a population... --> Romania (RO) has a population of.., Romania (RO/ROU) has a population of.., 
4. Also, to make it more accurate - while making the substitution, you could check if the words you are adding aren't already present in the sentence. eg. The UK, having capital London, is largely populated -->  The UK, with capital London, having capital London is largely populated. 
5. Also, please make a section in the README for Data and code provenance

Overall, nice work!

**Comment by pvf2005** [comment]
The suggestions from @kaustubhdhole were very interesting (especially with regard to the neighboring countries) and are now implemented. Additionally I added patterns for the spoken language(s) available in a country and the code is now easier to read (each pattern has its own method).

File: transformations/geonames_transformation/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+```Mangalia, a city in Romania, with a population of 39,619 inhabitants, is situated on the shores of the Black Sea```
+
+```Mangalia, a city in Europe, is situated on the shores of the Black Sea```
+
+```Mangalia, a city in Europe, with a population of 39,619 inhabitants, is situated on the shores of the Black Sea```
+
+
+## Data and code provenance
+This transformation makes use of city and country information obtained from GeoNames ( http://download.geonames.org/export/dump/ ).
```
Comment (kaustubhdhole) [inline transformations/geonames_transformation/README.md:38]: Also mention the license of Geonames: https://github.com/google/BIG-bench/blob/83a7a0dc23f41b13daeffc84f5bc8346bffd7163/bigbench/benchmark_tasks/spelling_bee/README.md

Comment (pvf2005) [reply transformations/geonames_transformation/README.md:38]: The license is now added

**Review by kaustubhdhole — APPROVED** [review]

**Review by ashish3586 — APPROVED** [review]

**Comment by kaustubhdhole** [comment]
Thank you very much for your changes. Your changes are merged. NLP has largely focussed on high-resource languages and domains. While this is not mandatory, if possible, we also encourage you to add a transformation in your mother tongue and/or a low-resource language which might need more attention from the NLP community. 

Thanks for your changes!

**Pull request merged**